src/storage-client/src/controller.rs

```diff
@@ -1833,6 +1753,233 @@ where
+        // leaked shards?
+
+        self.truncate_shards(&to_delete_shards).await;
+
+        let DurableCollectionMetadata {
+            remap_shard,
+            data_shard,
+        } = new_metadata;
+
+        // Update in memory collection metadata.
+        let metadata = self
+            .state
+            .collections
+            .get_mut(&id)
+            .expect("id {:?} must exist");
+        metadata.collection_metadata.data_shard = data_shard;
+        metadata.collection_metadata.remap_shard = remap_shard;
+    }
+
+    async fn truncate_shards(&mut self, shards: &[(ShardId, String)]) {
+        // Open a persist client to delete unused shards.
+        let persist_client = self
+            .persist
+            .lock()
+            .await
+            .open(self.persist_location.clone())
+            .await
+            .unwrap();
+
+        for (shard_id, shard_purpose) in shards {
+            let (mut write, mut read) = persist_client
+                .open::<crate::types::sources::data::SourceData, (), T, Diff>(
+                    *shard_id,
+                    shard_purpose.as_str(),
+                )
+                .await
+                .expect("invalid persist usage");
+
+            read.downgrade_since(&Antichain::new()).await;
+            write
+                .append(
+                    Vec::<((crate::types::sources::data::SourceData, ()), T, Diff)>::new(),
+                    write.upper().clone(),
+                    Antichain::new(),
+                )
+                .await
+                .expect("failed to connect")
+                .expect("failed to truncate write handle");
+
+            info!(
+                "successfully truncated shard's write handle for {:?}",
+                shard_id
+            );
+        }
+    }
+
+    /// For appropriate data sources, migrate the remap collection correlated
+    /// to `id` to one compatible with `Partitioned<PartitionId, MzOffset>`.
+    pub async fn migrate_remap_shard(&mut self, id: GlobalId, data_source: &DataSource) {
+        // We only want to migrate ingestion-based sources' remap collections
+        let partitioned_source = match data_source {
+            DataSource::Ingestion(IngestionDescription {
+                desc: crate::types::sources::SourceDesc { connection, .. },
+                ..
+            }) =>
+            // Only Kafka sources are partitioned
+            {
+                matches!(
+                    connection,
+                    crate::types::sources::GenericSourceConnection::Kafka(_)
+                )
+            }
+            _ => return,
+        };
+
+        let metadata = self
+            .state
+            .collections
+            .get(&id)
+            .expect("must have seen metadata inserted")
+            .collection_metadata
+            .clone();
+
+        let migrate_to_shard = ShardId::new();
+
+        if let None = remap_migration::RemapHandleMigrator::<T>::migrate(
```

Comment:
This function should probably return a `bool` as the `Some(())` case is never used and it doesn't carry any information either

Reply:
I rewrote this to instead return the shard ID of the shard we migrated to or none if the migration's unnecessary. Lets us sanely propagate `None` through `?` in a few spots.